sql: Alter Column Set Data Type General

RichardJCai opened the pull request:
This PR contains support for the ALTER TABLE ... ALTER COLUMN  ... TYPE ... command.

To ensure this PR isn't too large, this PR only contains support for ALTER COLUMN TYPE for columns that are not part of indexes, do not own sequences, do not have constraints, and does not support using an expression. These features will come in future PRs.

Enable experimental ALTER COLUMN TYPE general using SET enable_experimental_alter_column_type_general = true;

Release note (sql change): Support for ALTER COLUMN TYPE online schema change for changing a column type's data.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/46933)
<!-- Reviewable:end -->

rohany commented on pkg/sql/alter_table.go:
```diff
@@ -906,11 +905,73 @@ func applyColumnMutation(
+			// Make shadow column for now
+
+			// Can concurrent schema changes be happening?
+			currentMutationID := tableDesc.ClusterVersion.NextMutationID
```
It seems fine to disable this right now, we'll just want to change this to an unimplemented with issue.

rohany commented on pkg/sql/alter_table.go:
```diff
@@ -906,11 +905,73 @@ func applyColumnMutation(
-		}
+
+		default: // try conversion
+			//1. Create a new computed column c' using c::t'
```
Fix some spacing in these comments.

rohany commented on pkg/sql/alter_table.go:
```diff
@@ -906,11 +905,73 @@ func applyColumnMutation(
-				"type conversion not yet implemented")
-		}
+
+		default: // try conversion
```
I'd pull this whole case out into a separate function or file

rohany commented on pkg/sql/alter_table.go:
```diff
@@ -906,11 +905,73 @@ func applyColumnMutation(
+				return err == nil
+			}
+
+			name := sqlbase.GenerateUniqueConstraintName(
```
maybe put in a new name that is clear that this is the column that is replacing the new column.

rohany commented on pkg/sql/alter_table.go:
```diff
@@ -906,11 +905,73 @@ func applyColumnMutation(
+				nameExists,
+			)
+
+			var newColComputeExpr = proto.String(
```
I think you want to build the expression using the expression AST rather than formatting into a raw string.

rohany commented on pkg/sql/alter_table.go:
```diff
@@ -906,11 +905,73 @@ func applyColumnMutation(
+					fmt.Sprintf("%s::%s", col.DefaultExprStr(), t.ToType.String()))
+			}
+
+			id := tableDesc.NextColumnID
```
I think you'll want to just let allocate IDs do this for you.

rohany commented on pkg/sql/alter_table.go:
```diff
@@ -906,11 +905,73 @@ func applyColumnMutation(
+				Nullable: col.Nullable,
+				DefaultExpr: newColDefaultExpr,
+				Hidden: true,
+				UsesSequenceIds: col.UsesSequenceIds,
```
This seems suspicious -- check the checks that drop column makes before deleting a column. That will tell you what sort of stuff a column depends on so we can go and update all of those references.

rohany commented on pkg/sql/backfill.go:
```diff
@@ -204,6 +204,8 @@ func (sc *SchemaChanger) runBackfill(ctx context.Context) error {
 				}
 			case *sqlbase.DescriptorMutation_PrimaryKeySwap:
 				// The backfiller doesn't need to do anything here.
+			case *sqlbase.DescriptorMutation_ComputedColumnSwap:
```
don't make a separate case for this, add it to the above case using a comma

rohany commented on pkg/sql/schema_changer.go:
```diff
@@ -958,6 +960,54 @@ func (sc *SchemaChanger) done(ctx context.Context) (*sqlbase.ImmutableTableDescr
 			}
+
+			//** Computed Col Swap **//
+			if computedColumnSwap := mutation.GetComputedColumnSwap(); computedColumnSwap != nil {
```
this code seems duplicated with the code that starts a job for the primary key swap case. Can you pull that code out and reuse it here?

rohany commented on pkg/sql/sqlbase/structured.go:
```diff
@@ -2441,7 +2447,66 @@ func (desc *MutableTableDescriptor) RemoveColumnFromFamily(colID ColumnID) {
+func (desc *MutableTableDescriptor) RenameColumn(
+	col *ColumnDescriptor, newColName string,
+) error {
+	preFn := func(expr tree.Expr) (recurse bool, newExpr tree.Expr, err error) {
```
Id add some comments here, it wasn't clear on a first read what was going on.

rohany commented on pkg/sql/sqlbase/structured.go:
```diff
@@ -3091,6 +3156,82 @@ func (desc *MutableTableDescriptor) MakeMutationComplete(m DescriptorMutation) e
+				return err
+			}
+
+			newCol.Hidden = false
```
I dont think you want to be touching hidden here. Hidden is like for the rowid implicit column.

rohany commented on pkg/sql/sqlbase/structured.go:
```diff
@@ -3091,6 +3156,82 @@ func (desc *MutableTableDescriptor) MakeMutationComplete(m DescriptorMutation) e
+			newColName := newCol.Name
+			desc.RenameColumn(oldCol, uniqueName)
+			desc.RenameColumn(newCol, oldColName)
+			desc.RenameColumn(oldCol, newColName)
```
why do you need to do this

rohany commented on pkg/sql/sqlbase/structured.go:
```diff
@@ -3091,6 +3156,82 @@ func (desc *MutableTableDescriptor) MakeMutationComplete(m DescriptorMutation) e
+			// Handle indexes as well.
+
+			// Update Column Families, new Col is not in a family.
+			//for i := range desc.Families {
```
I dont think you need any of this -- you just have to make sure to add the newCol to the family of the old column, and put it in the position of the old column.

rohany commented on pkg/sql/sqlbase/structured.go:
```diff
@@ -3239,6 +3380,12 @@ func (desc *MutableTableDescriptor) AddPrimaryKeySwapMutation(swap *PrimaryKeySw
 }
 
+// AddComputedColumnSwapMutation adds a ComputedColumnSwap mutation to the table descriptor.
+func (desc *MutableTableDescriptor) AddComputedColumnSwapMutation(swap *ComputedColumnSwap) {
```
I think theres a missing case here for reverseMutations, which is called when a schema change job runs into an error or is cancelled.

rohany commented on pkg/sql/sqlbase/structured.proto:
```diff
@@ -972,3 +973,19 @@ message Descriptor {
+  optional uint32 new_column_id = 1 [(gogoproto.nullable) = false, (gogoproto.casttype) = "ColumnID"];
+//  // old_indexes and new_indexes are lists of IndexID's where the i'th index in old_indexes will be
+//  // swapped out with the i'th index in new_indexes.
+//  repeated uint32 old_indexes = 2 [(gogoproto.casttype) = "IndexID"];
```
delete these for now until they are added.

RichardJCai committed:
sql: add experimental support for non-trivial ALTER COLUMN TYPE.

Support experimental ALTER COLUMN TYPE general online schema change.
Enable experimental ALTER COLUMN TYPE general using SET enable_experimental_alter_column_type_general = true;

Release note (sql change): Added experimental support for ALTER COLUMN TYPE online schema change for changing a column type's data.

rohany commented on pkg/sql/alter_column_type.go:
```diff
@@ -0,0 +1,150 @@
+	case schemachange.ColumnConversionTrivial:
+		col.Type = *typ
+	case schemachange.ColumnConversionGeneral:
+		currentMutationID := tableDesc.ClusterVersion.NextMutationID
```
Can you pull this case out into a new function?

rohany commented on pkg/sql/alter_column_type.go:
```diff
@@ -0,0 +1,150 @@
+		return fmt.Errorf("unknown conversion for %s -> %s",
+			col.Type.SQLString(), typ.SQLString())
+	}
+
```
Can you call createOrUpdateSchemaChangeJob in here rather than letting the caller do it?

rohany commented on pkg/sql/sqlbase/structured.proto:
```diff
@@ -980,3 +983,17 @@ message Descriptor {
+// Column a while Column a becomes computed using a'.
+message ComputedColumnSwap {
+  option (gogoproto.equal) = true;
+  // old_column_index_id is the ID of the old column (column a).
```
These comments are out of date.

rohany commented on pkg/sql/sqlbase/structured.proto:
```diff
@@ -980,3 +983,17 @@ message Descriptor {
+message ComputedColumnSwap {
+  option (gogoproto.equal) = true;
+  // old_column_index_id is the ID of the old column (column a).
+  optional uint32 old_column_id = 4 [(gogoproto.nullable) = false, (gogoproto.casttype) = "ColumnID"];
```
Use proto fields 1 and 2 here

rohany commented on pkg/sql/sqlbase/structured.proto:
```diff
@@ -137,8 +137,10 @@ message ColumnDescriptor {
   optional uint32 logical_id = 13 [(gogoproto.nullable) = false,
    (gogoproto.customname) = "LogicalColumnID", (gogoproto.casttype) = "ColumnID"];
+  // Used to indicate column is used and dropped for ALTER COLUMN TYPE mutation.
+  optional bool alter_column_type_drop_column = 14 [(gogoproto.nullable) = false];
```
I'd name this something like "alter column type in progress"

rohany commented on pkg/sql/sqlbase/structured.go:
```diff
@@ -3122,6 +3189,84 @@ func (desc *MutableTableDescriptor) MakeMutationComplete(m DescriptorMutation) e
 				}
 			}
+		case *DescriptorMutation_ComputedColumnSwap:
+			args := t.ComputedColumnSwap
```
maybe pulling this into a separate function would be cleaner

rohany commented on pkg/sql/sqlbase/structured.go:
```diff
@@ -3122,6 +3189,84 @@ func (desc *MutableTableDescriptor) MakeMutationComplete(m DescriptorMutation) e
+
+			newCol.ComputeExpr = nil
+
+			unresolvedName := func(name string) *tree.UnresolvedName {
```
tree.MakeUnresolvedName performs this logic

rohany commented on pkg/sql/sqlbase/structured.go:
```diff
@@ -3122,6 +3189,84 @@ func (desc *MutableTableDescriptor) MakeMutationComplete(m DescriptorMutation) e
+				}
+			}
+
+			// oldCol should still have new values written to newCol in case
```
[nit]: Rephrase this as "oldCol still needs to have values written to it in case nodes read from it with a table descriptor from before the swap"

rohany commented on pkg/sql/sqlbase/structured.go:
```diff
@@ -3122,6 +3189,84 @@ func (desc *MutableTableDescriptor) MakeMutationComplete(m DescriptorMutation) e
+			// oldCol should still have new values written to newCol in case
+			// it has a read from previous version.
+
+			oldColComputeExpr := tree.CastExpr{Expr: unresolvedName(oldCol.Name), Type: oldCol.DatumType(), SyntaxMode: tree.CastShort}
```
[nit] Format this as 
``` 
CastExpr{
  Type..
   Expr..
...
}

rohany commented on pkg/sql/sqlbase/structured.go:
```diff
@@ -3122,6 +3189,84 @@ func (desc *MutableTableDescriptor) MakeMutationComplete(m DescriptorMutation) e
+			s := tree.Serialize(&oldColComputeExpr)
+			oldCol.ComputeExpr = &s
+
+			//Rename newCol to oldCol.Name.
```
[nit]: comments missing a space

rohany commented on pkg/sql/sqlbase/structured.go:
```diff
@@ -3122,6 +3189,84 @@ func (desc *MutableTableDescriptor) MakeMutationComplete(m DescriptorMutation) e
+
+			// Swap Column Family references.
+			// This preserves the ordering of column families when querying for it.
+			for i := range desc.Families {
```
Can this be a separate function?

rohany commented on pkg/sql/sqlbase/structured.go:
```diff
@@ -3122,6 +3189,84 @@ func (desc *MutableTableDescriptor) MakeMutationComplete(m DescriptorMutation) e
+			}
+
+			// Swap Column Orders.
+			*newCol, *oldCol = *oldCol, *newCol
```
Why do you need to swap them if you are just going to remove oldCol from the columns list anyway?

RichardJCai replied:
Really good point :P

rohany commented on pkg/sql/sqlbase/structured.go:
```diff
@@ -3122,6 +3189,84 @@ func (desc *MutableTableDescriptor) MakeMutationComplete(m DescriptorMutation) e
+			for i := range desc.Columns {
+				if desc.Columns[i].ID == oldCol.ID {
+					desc.AddColumnMutation(oldCol, DescriptorMutation_DROP)
+					// Use [:i:i] to prevent reuse of existing slice, or outstanding refs
```
Do you need to do this? What goes wrong when you dont?

RichardJCai replied:
I'm not sure actually, I took this from the drop column code.

rohany commented on pkg/sql/schema_changer.go:
```diff
@@ -1384,6 +1363,10 @@ func (sc *SchemaChanger) reverseMutation(
 		if pkSwap := mutation.GetPrimaryKeySwap(); pkSwap != nil {
 			return mutation, columns
 		}
+		if computedColumnsSwap := mutation.GetColumn(); computedColumnsSwap != nil {
```
Update the comment above to try and group these two cases together

rohany commented on pkg/sql/logictest/testdata/logic_test/alter_column_type:
```diff
@@ -161,3 +161,86 @@ ALTER TABLE t ALTER s TYPE STRING
 statement ok
 DROP TABLE t
+
+# Test cases for ALTER COLUMN TYPE requiring a cast
```
end with a period

rohany commented on pkg/sql/logictest/testdata/logic_test/alter_column_type:
```diff
@@ -161,3 +161,86 @@ ALTER TABLE t ALTER s TYPE STRING
+
+# Test cases for ALTER COLUMN TYPE requiring a cast
+
+# Failed alter
```
Use complete sentences

rohany commented on pkg/sql/logictest/testdata/logic_test/alter_column_type:
```diff
@@ -161,3 +161,86 @@ ALTER TABLE t ALTER s TYPE STRING
+statement error pq: parsing as type timestamp: could not parse "hello"
+ALTER TABLE t1 ALTER COLUMN date TYPE timestamp
+
+subtest ColumnNotPartOfIndex
```
[nit]: I think we snake cases these subtest names, but I'm not sure.

rohany commented on pkg/sql/logictest/testdata/logic_test/alter_column_type:
```diff
@@ -161,3 +161,86 @@ ALTER TABLE t ALTER s TYPE STRING
+INSERT INTO t2 VALUES ('5')
+
+statement error pq: value type int doesn't match type string of column "id"
+INSERT INTO t2 VALUES (6)
```
Missing a test here that the error is what you expect when you insert something that can't be converted back to an int

RichardJCai replied:
I added that as a test in alter_column_type_test.go, `TestInsertBeforeOldColumnIsDropped`.

My concern for not having it as a logic test is it could possibly be flakey if the garbage collector somehow drops the column quickly - this won't be the case now but possible if the gc changes.

rohany commented on pkg/sql/logictest/testdata/logic_test/alter_column_type:
```diff
@@ -161,3 +161,86 @@ ALTER TABLE t ALTER s TYPE STRING
+statement ok
+ALTER TABLE d.t1 ALTER COLUMN y TYPE STRING
+
+query TTBTTTB
```
Need a rowsort here

RichardJCai replied:
Show Columns has a baked in ORDER BY ordinal_position so I think a rowsort shouldn't be necessary.

rohany commented on pkg/sql/insert_fast_path.go:
```diff
@@ -282,6 +283,14 @@ func (n *insertFastPathNode) BatchedNext(params runParams) (bool, error) {
 			inputRow[col], err = typedExpr.Eval(params.EvalContext())
 			if err != nil {
+				// Intercept parse error due to ALTER COLUMN TYPE schema change.
+				if n.run.insertRun.insertCols[col].AlterColumnTypeDropColumn {
```
I think that you also want to put this error catch in the insert non-fast path node, and the updater.  It would be good to wrap this in a function for easy reuse

rohany commented on pkg/sql/insert_fast_path.go:
```diff
@@ -282,6 +283,14 @@ func (n *insertFastPathNode) BatchedNext(params runParams) (bool, error) {
+				// Intercept parse error due to ALTER COLUMN TYPE schema change.
+				if n.run.insertRun.insertCols[col].AlterColumnTypeDropColumn {
+					code := pgerror.GetPGCode(err)
+					if code == "22P02" {
```
Is there a constant you can compare against here?

rohany commented on pkg/sql/insert_fast_path.go:
```diff
@@ -282,6 +283,14 @@ func (n *insertFastPathNode) BatchedNext(params runParams) (bool, error) {
+				if n.run.insertRun.insertCols[col].AlterColumnTypeDropColumn {
+					code := pgerror.GetPGCode(err)
+					if code == "22P02" {
+						return false, errors.Wrapf(err,
```
I'd make this error a bit more general -- something like "this column is still finalizing the alter column set datatype schema change ..."

rohany commented on pkg/sql/alter_column_type_test.go:
```diff
@@ -0,0 +1,223 @@
+
+	blockGC := make(chan struct{})
+	params.Knobs = base.TestingKnobs{
+		GCJob: &sql.GCJobTestingKnobs{RunBeforeResume: func(_ int64) error { <-blockGC; return nil }},
```
I don't think the GC job is the right one to block -- you want to block the schema change job adopter.

rohany commented on pkg/sql/alter_column_type_test.go:
```diff
@@ -0,0 +1,223 @@
+	params.Knobs = base.TestingKnobs{
+		SQLSchemaChanger: &sql.SchemaChangerTestingKnobs{
+			RunBeforeComputedColumnSwap: func() {
+				// Notify the tester that the primary key swap is about to happen.
```
outdated comment

rohany commented on pkg/sql/alter_column_type_test.go:
```diff
@@ -0,0 +1,223 @@
+}
+
+// Column Families test
+func TestColumnFamilies(t *testing.T) {
```
This could be a logictest

rohany commented on pkg/sql/alter_column_type_test.go:
```diff
@@ -0,0 +1,223 @@
+	}
+}
+
+func TestAlterColumnTypeTimestampTZ(t *testing.T) {
```
This could be a logictest too

rohany commented on pkg/sql/alter_column_type_test.go:
```diff
@@ -0,0 +1,223 @@
+	if _, err := sqlDB.Exec(`
+CREATE DATABASE t;
+BEGIN;
+CREATE TABLE t.test (x INT);
```
We'd want a logic test for this too (that is verifying the output). The primary key change test for this case does a similar thing, can you combine them?

rohany commented on pkg/sql/alter_column_type.go:
```diff
@@ -0,0 +1,150 @@
+			NewColumnId: newCol.ID,
+		}
+
+		tableDesc.AddComputedColumnSwapMutation(swapArgs)
```
Can you add a notice about the insert behavior/potential errors?

rohany commented on pkg/sql/alter_column_type_test.go:
```diff
@@ -0,0 +1,223 @@
+	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
+	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
+)
+
```
Id like to see tests that:
* run this schema change while operations are going on
* Cancel and rollback the job

rohany commented on pkg/sql/logictest/testdata/logic_test/alter_column_type:
```diff
@@ -161,3 +161,86 @@ ALTER TABLE t ALTER s TYPE STRING
 
 statement ok
 DROP TABLE t
+
```
Add some tests for how this interacts with preceeding schema changes. Concurrent schema changes probably needs a schema_changer_test

rohany reviewed (changes requested):
Nice work!

It seems like you've included an update to yarn-vendor here while rebasing.

Some general comments/things I want to see added:
* Feature flags / variable settings to enable use of this case
* Version gating
* More testing

rohany commented:
I'm going to loop in the SQL Schema team as well.

rohany commented on pkg/sql/alter_column_type.go:
```diff
@@ -0,0 +1,225 @@
+	case schemachange.ColumnConversionTrivial:
+		col.Type = *typ
+	case schemachange.ColumnConversionGeneral:
+		// Make sure that all nodes in the cluster are able to perform general alter column type conversions.
```
[nit]: Can you add some spacing + comments below?

rohany commented on pkg/sql/alter_column_type.go:
```diff
@@ -0,0 +1,225 @@
+		if err := alterColumnTypeGeneral(tableDesc, col, t); err != nil {
+			return err
+		}
+		if err := params.p.createOrUpdateSchemaChangeJob(params.ctx, tableDesc, "alter column type", sqlbase.InvalidMutationID); err != nil {
```
We should have a real mutation ID to pass in here right?

rohany commented on pkg/sql/alter_column_type.go:
```diff
@@ -0,0 +1,225 @@
+		if err := params.p.createOrUpdateSchemaChangeJob(params.ctx, tableDesc, "alter column type", sqlbase.InvalidMutationID); err != nil {
+			return err
+		}
+		params.p.SendClientNotice(params.ctx, pgerror.Noticef("alter column type changes are finalized asynchronosuly;"+
```
spelling error in asynchronously

rohany commented on pkg/sql/alter_column_type.go:
```diff
@@ -0,0 +1,225 @@
+	t *tree.AlterTableAlterColumnType,
+) error {
+
+	usingExpressionNotSupportedErr := unimplemented.NewWithIssuef(
```
these can be allocated outside the function in a variable

rohany commented on pkg/sql/alter_column_type.go:
```diff
@@ -0,0 +1,225 @@
+	if t.Using != nil {
+		return usingExpressionNotSupportedErr
+	}
+
```
[nit]: Can you add some comments here?

rohany commented on pkg/sql/alter_column_type.go:
```diff
@@ -0,0 +1,225 @@
+		return err == nil
+	}
+
+	shadowColName := sqlbase.GenerateUniqueConstraintName(
```
[nit]: can this fit on one line

rohany commented on pkg/sql/alter_column_type.go:
```diff
@@ -0,0 +1,225 @@
+		nameExists,
+	)
+
+	// Compute Expr.
```
What does this comment mean?

rohany commented on pkg/sql/alter_column_type.go:
```diff
@@ -0,0 +1,225 @@
+	newColComputeExpr := &s
+
+	// May run into runtime error if DEFAULT value of column cannot be casted.
+	// In this case right now, the schema change is not rolled back.
```
I don't follow what this comment is saying

rohany commented on pkg/sql/alter_column_type.go:
```diff
@@ -0,0 +1,225 @@
+
+	// Compute Expr.
+	unresolvedName := func(name string) *tree.UnresolvedName {
+		return &tree.UnresolvedName{
```
Apparently the thing to do is make a `tree.ColumnItem` rather than an unresolved name

rohany commented on pkg/sql/alter_column_type.go:
```diff
@@ -0,0 +1,225 @@
+		Type:            *t.ToType,
+		Nullable:        col.Nullable,
+		DefaultExpr:     newColDefaultExpr,
+		UsesSequenceIds: col.UsesSequenceIds,
```
do you need to add this column to the backreferences of any sequences or anything? The drop column code checks everything that depends on a column.

rohany commented on pkg/sql/alter_column_type.go:
```diff
@@ -0,0 +1,225 @@
+
+	// Ensure shadow column is created in the same column family as the original
+	// so backfiller writes to the same column family.
+	var family sqlbase.ColumnFamilyDescriptor
```
Can you pull this out into a utility method on a table descriptor -- i feel like this is done in alot of places.

rohany commented on pkg/sql/alter_column_type_test.go:
```diff
@@ -0,0 +1,521 @@
+	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
+)
+
+func TestAlterColumnTypeInTxn(t *testing.T) {
```
This can just be a logictest right?

rohany commented on pkg/sql/alter_column_type_test.go:
```diff
@@ -0,0 +1,521 @@
+	params, _ := tests.CreateTestServerParams()
+
+	s, sqlDB, _ := serverutils.StartServer(t, params)
+	ctx := context.TODO()
```
[nit] you can just use context.Background() here

rohany commented on pkg/sql/alter_column_type_test.go:
```diff
@@ -0,0 +1,521 @@
+	}
+}
+
+// TestAlterColumnTypeFailureRollback tests running a primary key
```
out of date comment

rohany commented on pkg/sql/alter_column_type_test.go:
```diff
@@ -0,0 +1,521 @@
+	}
+}
+
+// Test while old column still exists - parse error.
```
incomplete sentence

rohany commented on pkg/sql/alter_column_type_test.go:
```diff
@@ -0,0 +1,521 @@
+	// Expect to find two drop mutations, one to drop the Column, one to drop the
+	// Computed Column swap.
+	if len(desc.Mutations) != 2 {
+		t.Fatalf("expected to find 2 mutations, but found %d", len(desc.Mutations))
```
incorrectly nested?

rohany commented on pkg/sql/alter_column_type_test.go:
```diff
@@ -0,0 +1,521 @@
+	}
+
+	desc := sqlbase.GetTableDescriptor(kvDB, "t", "test")
+	// Expect to find two drop mutations, one to drop the Column, one to drop the
```
I think a better test would be to continue polling the descriptor in a succeeds soon and ensure that we don't have any pending descriptors (i.e. the mutations get cleaned up).

rohany commented on pkg/sql/alter_column_type_test.go:
```diff
@@ -0,0 +1,521 @@
+	}
+}
+
+func TestAddDropColumnDuringAlterColumnType(t *testing.T) {
```
I dont understand what this test is specifically testing / why dropping a column is problematic?

rohany commented on pkg/sql/alter_column_type_test.go:
```diff
@@ -0,0 +1,521 @@
+	}
+}
+
+func TestIntToStringBasic(t *testing.T) {
```
This could definitely be a logictest

rohany commented on pkg/sql/alter_column_type_test.go:
```diff
@@ -0,0 +1,521 @@
+	}
+}
+
+func TestAlterColumnTypeBeforeAlterPrimaryKey(t *testing.T) {
```
I'm not sure what this test is adding

rohany commented on pkg/sql/alter_column_type_test.go:
```diff
@@ -0,0 +1,521 @@
+	}
+}
+
+func TestAlterPrimaryKeyBeforeAlterColumnType(t *testing.T) {
```
I'd rename this to something like TestSchemaChangeBeforeAlterColumnType -- theres nothing unique about the primary key change here right?

rohany commented on pkg/sql/alter_column_type_test.go:
```diff
@@ -0,0 +1,521 @@
+	}
+}
+
+func TestAlterColumnTypeWithAlterPrimaryKey(t *testing.T) {
```
This test seems like it just depends on alter column working and then alter primary key working -- what does it add?

rohany commented on pkg/sql/exec_util.go:
```diff
@@ -192,6 +192,12 @@ var insertFastPathClusterMode = settings.RegisterBoolSetting(
 )
 
+var alterColumnTypeGeneralMode = settings.RegisterBoolSetting(
+	"sql.defaults.alter_column_type.enabled",
```
this should be experimental for now

rohany commented on pkg/sql/insert_fast_path.go:
```diff
@@ -282,6 +283,14 @@ func (n *insertFastPathNode) BatchedNext(params runParams) (bool, error) {
+					code := pgerror.GetPGCode(err)
+					if code == "22P02" {
+						return false, errors.Wrapf(err,
+							"Column is under ALTER COLUMN TYPE schema change, this insert may not be supported until the schema change is finished (the original column must be dropped)")
```
change to something like "column 'name' is still undergoing the ALTER COLUMN TYPE schema change..." The insert has already been rejected, why say "may"?

rohany commented on pkg/sql/logictest/logic.go:
```diff
@@ -1532,7 +1532,6 @@ func (t *logicTest) processSubtest(
 				)
 			}
 			repeat = count
-
```
[nit]: no need to remove this

rohany commented on pkg/sql/logictest/testdata/logic_test/alter_column_type:
```diff
@@ -161,3 +161,170 @@ ALTER TABLE t ALTER s TYPE STRING
+4  4  4
+
+statement ok
+CREATE DATABASE d
```
is there a reason to make a new db

rohany commented on pkg/sql/logictest/testdata/logic_test/alter_table_mixed_19.2_20.1:
```diff
@@ -171,7 +171,7 @@ CREATE TABLE regression_47110(a TIME)
 ALTER TABLE regression_47110 ADD COLUMN b TIMETZ
 
-statement error type TIMETZ is not supported until version upgrade is finalized
+statement error all nodes are not the correct version for alter column type general
```
Is this error message correct to be changed

RichardJCai replied:
I think what happens here is ALTER COLUMN TYPE is disabled for general type conversions including this one until all versions are upgraded - this error appears before we actually use the type TIMETZ so we don't reach that error.

rohany replied:
I'd just delete this test then.

rohany commented on pkg/sql/schema_changer.go:
```diff
@@ -1772,3 +1752,52 @@ func init() {
 	jobs.RegisterConstructor(jobspb.TypeSchemaChange, createResumerFn)
 }
+
+func (sc *SchemaChanger) queueCleanupJobs(
```
comment what this function does

rohany commented on pkg/sql/sqlbase/structured.go:
```diff
@@ -889,11 +892,16 @@ func (desc *TableDescriptor) MaybeFillInDescriptor(
 	desc.maybeUpgradeFormatVersion()
 	desc.Privileges.MaybeFixPrivileges(desc.ID)
+
+	var err error
```
did anything change here and below?

rohany commented on pkg/sql/sqlbase/structured.go:
```diff
@@ -3097,6 +3226,84 @@ func (desc *MutableTableDescriptor) MakeMutationComplete(m DescriptorMutation) e
+			// Swap Column Family references.
+			// This preserves the ordering of column families when querying for it.
+			for i := range desc.Families {
+				for j := range desc.Families[i].ColumnIDs {
```
We decided this was incorrect right?

rohany commented on pkg/sql/testutils.go:
```diff
@@ -126,3 +129,36 @@ func (dsp *DistSQLPlanner) Exec(
+//			expected := [][]interface{}{{1, "1", 1}, {2, "2", 2}, {3, "HELLO", 3}}
+// dest: pointers to the desired types for the Scan to read into.
+//			for the example above, dest should contain *int, *string, *int.
+func QueryExpect(
```
we must have something that does this already? Either way, i think wherever this is used a combination of the normal query + require.Equal() would suffice.

rohany reviewed (changes requested):
Looking better! Mostly just comments about formatting and comments.

lucy-zhang commented:
Heads up, I commented on the RFC because I wanted to clarify some things before taking a closer look at this PR.

rohany commented on pkg/sql/alter_column_type.go:
```diff
@@ -0,0 +1,239 @@
+			s := tree.Serialize(tree.DNull)
+			newColDefaultExpr = &s
+		} else {
+			// The "default" default expression for the new column is casting the
```
what does the second default mean here?

RichardJCai replied:
I meant the default expression for the column by default is casting the old default column to the new type. May be different if the user provides an expression for the schema change

rohany commented on pkg/sql/alter_column_type.go:
```diff
@@ -0,0 +1,239 @@
+		}
+	}
+
+	id := tableDesc.NextColumnID
```
Whats the reason again that we aren't using allocate ID's?

RichardJCai replied:
No reason, I think it just jumped out to me to specify every field in the ColumnDescriptor but setting ID to 0 and calling AllocateIDs should be okay.

rohany commented on pkg/sql/alter_column_type.go:
```diff
@@ -0,0 +1,239 @@
+	}
+
+	tableDesc.AddColumnMutation(&newCol, sqlbase.DescriptorMutation_ADD)
+	for _, sequenceID := range col.OwnsSequenceIds {
```
As discussed, lets deal with columns that have dependencies later.

rohany commented on pkg/sql/information_schema.go:
```diff
@@ -487,6 +487,9 @@ func characterMaximumLength(colType *types.T) tree.Datum {
 			if colType.Width() > 0 {
 				return colType.Width(), true
 			}
+			if colType.Width() > 0 {
```
merge conflict?

rohany commented on pkg/sql/insert.go:
```diff
@@ -223,6 +226,21 @@ func (n *insertNode) BatchedNext(params runParams) (bool, error) {
+				// parse error is specifically from the column undergoing the
+				// alter column type schema change.
+
+				// Intercept parse error due to ALTER COLUMN TYPE schema change.
```
Yeah, I don't think we can safely do this here... Not sure how to catch the error here.

RichardJCai replied:
Jordan said it should be okay to say one of the columns is undergoing alter column type and say it MAY be because of this. The argument is that the user should expect behaviour like this for the online alter column type schema change.

rohany replied:
OK, that seems reasonable.

We should use the method that intercepts the error here though.

rohany commented on pkg/sql/insert_fast_path.go:
```diff
@@ -282,6 +284,16 @@ func (n *insertFastPathNode) BatchedNext(params runParams) (bool, error) {
 			inputRow[col], err = typedExpr.Eval(params.EvalContext())
 			if err != nil {
+				// Intercept parse error due to ALTER COLUMN TYPE schema change.
+				insertCol := n.run.insertRun.insertCols[col]
```
Pull this into a helper method.

rohany commented on pkg/sql/logical_column_id_test.go:
```diff
@@ -0,0 +1,76 @@
+// Copyright 2020 The Cockroach Authors.
```
merge conflict?

rohany commented on pkg/sql/logictest/testdata/logic_test/alter_column_type:
```diff
@@ -161,3 +161,210 @@ ALTER TABLE t ALTER s TYPE STRING
+BEGIN;
+CREATE TABLE t14 (x INT);
+ALTER TABLE t14 ALTER COLUMN x TYPE STRING;
+COMMIT;
```
The file is empty, so I'm adding a comment here -- what is the alter_column_type_other file?

rohany commented on pkg/sql/schema_changer.go:
```diff
@@ -1771,3 +1767,51 @@ func init() {
+		if err != nil {
+			return nil, err
+		}
+		log.VEventf(ctx, 2, "created job %d to drop previous indexes", *job.ID())
```
this log is out of date for your case -- should it be pulled out of this method, or the message should be generalized

rohany commented on pkg/sql/exec_util.go:
```diff
@@ -224,6 +224,12 @@ var insertFastPathClusterMode = settings.RegisterBoolSetting(
 	true,
 )
 
+var alterColumnTypeGeneralMode = settings.RegisterBoolSetting(
```
this variable should be prefaced with experimental as well

rohany commented on pkg/sql/sqlbase/structured.go:
```diff
@@ -387,6 +387,9 @@ func GetTableDescFromIDWithFKsChanged(
 		return nil, false, err
 	}
+
+	table.MaybeFillInLogicalColumnID()
```
merge conflict?

rohany commented on pkg/sql/sqlbase/structured.go:
```diff
@@ -889,11 +892,15 @@ func (desc *TableDescriptor) MaybeFillInDescriptor(
 		}
 	}
+
+	desc.MaybeFillInLogicalColumnID()
```
merge conflicts?

rohany commented on pkg/sql/sqlbase/structured.go:
```diff
@@ -1198,6 +1205,27 @@ func (desc *MutableTableDescriptor) MaybeFillColumnID(
+// MaybeFillInLogicalColumnID assigns a LogicalColumnID to a column
+// if the ColumnID is set and the LogicalColumnID is not
+// it assigns the ColumnID as the LogicalColumnID.
+func (desc *TableDescriptor) MaybeFillInLogicalColumnID() {
```
merge conflicts?

rohany commented on pkg/sql/sqlbase/structured.go:
```diff
@@ -3129,6 +3262,97 @@ func (desc *MutableTableDescriptor) MakeMutationComplete(m DescriptorMutation) e
+	// ordering of ColumnIDs in the family.
+	// The old column will be removed from column families
+	// when the column is dropped.
+	for i := range desc.Families {
```
We agreed that changing the families like this was incorrect!

rohany commented on pkg/sql/sqlbase/structured.go:
```diff
@@ -3129,6 +3262,97 @@ func (desc *MutableTableDescriptor) MakeMutationComplete(m DescriptorMutation) e
+
+	// Now move newCol into oldCol's position in the descriptor slice
+	// and remove the original entry for newCol in the descriptor slice.
+	for i := range desc.Columns {
```
I remember discussing that we might want to first remove new column from the list of columns at the beginning of the method for clarity

rohany commented on pkg/sql/sqlbase/structured.go:
```diff
@@ -3863,6 +4093,20 @@ func (desc *ColumnDescriptor) CheckCanBeFKRef() error {
 
+// GetColumnFamily returns the ColumnFamilyDescriptor for the
+// the family the column is part of.
+func (desc *TableDescriptor) GetColumnFamily(colID ColumnID) (*ColumnFamilyDescriptor, error) {
```
better name: GetFamilyOfColumn

rohany commented on pkg/sql/sqlbase/system.go:
```diff
@@ -378,9 +378,9 @@ var (
-			{Name: "parentID", ID: 1, Type: *types.Int},
-			{Name: "name", ID: 2, Type: *types.String},
-			{Name: "id", ID: 3, Type: *types.Int, Nullable: true},
+			{Name: "parentID", ID: 1, LogicalColumnID: 1, Type: *types.Int},
```
This whole file is one large merge conflict?

rohany reviewed (changes requested):
I haven't read the tests yet, but here are some things to clean up

RichardJCai commented on pkg/sql/alter_column_type.go:
```diff
@@ -0,0 +1,239 @@
+		col.Type = *typ
+	case schemachange.ColumnConversionGeneral:
+		// Make sure that all nodes in the cluster are able to perform general alter column type conversions.
+		if !params.p.ExecCfg().Settings.Version.IsActive(params.ctx, clusterversion.VersionAlterColumnTypeGeneral) {
```
Wasn't passing params into alterColumnTypeGeneral when I wrote these checks but should have moved these checks inside once I passed params in.

rohany commented on pkg/sql/alter_column_type.go:
```diff
@@ -0,0 +1,231 @@
+
+	// Disallow ALTER COLUMN TYPE general for columns that own sequences.
+	if len(col.OwnsSequenceIds) != 0 {
+		fmt.Println(col.OwnsSequenceIds)
```
print

rohany commented on pkg/sql/sqlbase/structured.go:
```diff
@@ -3129,6 +3236,115 @@ func (desc *MutableTableDescriptor) MakeMutationComplete(m DescriptorMutation) e
+	s := tree.Serialize(&oldColComputeExpr)
+	oldCol.ComputeExpr = &s
+
+	//Generate unique name for old column.
```
[nit] space after the comment

rohany commented on pkg/sql/sqlbase/structured.go:
```diff
@@ -3129,6 +3236,115 @@ func (desc *MutableTableDescriptor) MakeMutationComplete(m DescriptorMutation) e
+
+	// Remove ownership for sequences for the old column otherwise
+	// the owned sequences will be dropped when the old column is dropped.
+	oldCol.OwnsSequenceIds = nil
```
Since we aren't supporting sequences right now, this shouldn't be here.

rohany commented on pkg/sql/sqlbase/structured.proto:
```diff
@@ -985,3 +988,13 @@ message Descriptor {
+// ComputedColumnSwap is a mutation corresponding to the atomic swap phase
+// where Column a' that is computed using Column a is swapped to replace
+// Column a while Column a becomes computed using a'.
+message ComputedColumnSwap {
```
nit, but can you move this near the other mutations?

rohany commented on pkg/sql/logictest/testdata/logic_test/alter_column_type:
```diff
@@ -161,3 +161,197 @@ ALTER TABLE t ALTER s TYPE STRING
+statement error pq: parsing as type timestamp: could not parse "hello"
+ALTER TABLE t1 ALTER COLUMN date TYPE timestamp
+
+#ColumnNotPartOfIndex
```
[nit] comments should be complete sentences

rohany commented on pkg/sql/logictest/testdata/logic_test/alter_column_type:
```diff
@@ -161,3 +161,197 @@ ALTER TABLE t ALTER s TYPE STRING
+
+subtest alter_column_type_general
+
+statement ok
```
State above each test what you are trying test -- it makes the tests easier to follow.

rohany commented on pkg/sql/logictest/testdata/logic_test/alter_column_type:
```diff
@@ -161,3 +161,197 @@ ALTER TABLE t ALTER s TYPE STRING
+2016-01-26 16:10:10.55 +0100 CET
+2016-01-26 16:10:10.556 +0100 CET
+
+# ensure column families stay the same
```
[nit] capital letters and periods

rohany commented on pkg/sql/logictest/testdata/logic_test/alter_column_type:
```diff
@@ -161,3 +161,197 @@ ALTER TABLE t ALTER s TYPE STRING
+1970-01-02 00:00:00 +0000 +0000  4
+
+# ensure a runtime error rollsback correctly.
+statement ok
```
Is something missing here?

rohany commented on pkg/sql/logictest/testdata/logic_test/alter_column_type:
```diff
@@ -161,3 +161,197 @@ ALTER TABLE t ALTER s TYPE STRING
+1970-01-02 00:00:00 +0000 +0000  3
+1970-01-02 00:00:00 +0000 +0000  4
+
+# ensure a runtime error rollsback correctly.
```
periods and capitalization here please

rohany commented on pkg/sql/logictest/testdata/logic_test/alter_column_type:
```diff
@@ -161,3 +161,197 @@ ALTER TABLE t ALTER s TYPE STRING
+statement error pq: unimplemented: alter column type using expression is not supported\nHINT: You have attempted to use a feature that is not yet implemented.\nSee: https://github.com/cockroachdb/cockroach/issues/47706
+ALTER TABLE t11 ALTER COLUMN x TYPE STRING USING CAST(x AS STRING)
+
+# dropping a column that owns a sequence should not be allowed.
```
ditto

rohany commented on pkg/sql/logictest/testdata/logic_test/alter_column_type:
```diff
@@ -161,3 +161,197 @@ ALTER TABLE t ALTER s TYPE STRING
+statement error pq: unimplemented: alter column type using expression is not supported\nHINT: You have attempted to use a feature that is not yet implemented.\nSee: https://github.com/cockroachdb/cockroach/issues/47706
+ALTER TABLE t11 ALTER COLUMN x TYPE STRING USING CAST(x AS STRING)
+
+# dropping a column that owns a sequence should not be allowed.
```
On this note, can you see if we need to disable column type changes on columns that are used in views?

RichardJCai replied:
Tested views, they should be okay, they use the table id and not column id and the selects work as expected with the correct type.

rohany commented on pkg/sql/alter_column_type_test.go:
```diff
@@ -0,0 +1,232 @@
+INSERT INTO test VALUES ('hello');
+`)
+
+	expected = "pq: This table is still undergoing the ALTER COLUMN TYPE schema change, this insert may not be supported until the schema change is finalized: could not parse \"hello\" as type int: strconv.ParseInt: parsing \"hello\": invalid syntax"
```
err is the same here right? you don't need to redefine it.

RichardJCai replied:
The error is slightly different for the fast path insert vs non fast path insert. I put `may` be disabled due to ALTER COLUMN TYPE schema change for the non fast path insert because it could be an actual parse error - due to not actually being able to check the column.

rohany replied:
I see -- is it possible to just have this test search for the prefix of "this table is still undergoing ..."? I don't think there's a need for the whole string.

rohany commented on pkg/sql/alter_column_type_test.go:
```diff
@@ -0,0 +1,232 @@
+	sqlDB.CheckQueryResults(t, "SHOW CREATE TABLE t.test", expected)
+}
+
+// TestAlterColumnTypeFailureRollback tests running a alter column type
```
I think this comment is out of date?

rohany reviewed (changes requested):
Some more comments on the test code -- sorry this PR large, so I keep finding things on later passes.

rohany commented on pkg/sql/insert_fast_path.go:
```diff
@@ -343,3 +346,43 @@ func (n *insertFastPathNode) Close(ctx context.Context) {
+// If colNum is not -1, only the colNum'th column in insertCols will be checked
+// for AlterColumnTypeInProgress, otherwise every column in insertCols will
+// be checked.
+func interceptAlterColumnTypeParseError(
```
I feel like all the duplicated nesting this function does could be deduplicated

RichardJCai replied:
Made some changes to this function to deduplicate code, let me know what you think.

rohany commented on pkg/sql/schema_changer.go:
```diff
@@ -1782,3 +1769,50 @@ func init() {
+
+// queueCleanupJobs checks for columns and indexes that have been dropped
+// and queues a job for the garbage collector to delete the data.
+func (sc *SchemaChanger) queueCleanupJobs(
```
I'd change this comment to something like "... checks if the completed schema change needs to start a child job to clean up dropped schema elements."

rohany commented on pkg/sql/schema_changer.go:
```diff
@@ -1482,6 +1464,11 @@ type SchemaChangerTestingKnobs struct {
+	// SkipComputedColumnSwapCleanup determines if the cleanup job should be ran for
+	// the computed column swap. If the cleanup job is not run, the old columns
+	// and indexes will not be cleaned up.
+	SkipComputedColumnSwapCleanup bool
```
This testing knob feels pretty awkward -- is it possible to do something with the backfill notification like in the other tests for the test you are using this in?

RichardJCai replied:
I tried the backfill knobs and but I don't think it supports this case - need to let the schema change "finish" but not let child jobs run. I added a new RunBeforeChildJobs knob which I think aligns more with the existing knobs, let me know if you think this makes sense.

rohany commented:
cc @lucy-zhang can you give this PR a review?

lucy-zhang commented:
Yeah, for sure. Do both of you think we should also try to get more sign-offs on the RFC, and ideally merge it? I think we're at a place where we all basically agree on this implementation, but it'd be good to make that more explicit (and figure out if there are any lingering questions we need to answer right now).

RichardJCai commented:
> Yeah, for sure. Do both of you think we should also try to get more sign-offs on the RFC, and ideally merge it? I think we're at a place where we all basically agree on this implementation, but it'd be good to make that more explicit (and figure out if there are any lingering questions we need to answer right now).

I think merging the RFC would be a good idea, although how many sign-offs do we need/want?

lucy-zhang commented:
I followed up on the PR for the RFC.

RichardJCai commented:
> I followed up on the PR for the RFC.

Now that the RFC is merged, can you review this when you have some time?

jordanlewis reviewed (commented):
nit: commit messages and PRs should have a package prefix and have a summary of the change. Example:

```
sql: add experimental support for non-trivial ALTER COLUMN TYPE
```

> Dismissed @rohany from 110 discussions.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46933#-:-M7ers2H0r_LgpMsXTes:b-7kfay5)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @lucy-zhang, and @RichardJCai)

___
*[pkg/sql/alter_column_type.go, line 131 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/46933#-M7etDMsB8uJZUrRN8_A:-M7etDMsB8uJZUrRN8_B:bybmm3z) ([raw file](https://github.com/cockroachdb/cockroach/blob/7df49a0ffbaba8886de54804adcf69e939c14ba9/pkg/sql/alter_column_type.go#L131)):*
> ```Go
> 	if !params.p.ExecCfg().Settings.Version.IsActive(params.ctx, clusterversion.VersionAlterColumnTypeGeneral) {
> 		return pgerror.Newf(pgcode.FeatureNotSupported,
> 			"all nodes are not the correct version for alter column type general")
> ```

nit: "version %v must be finalized to run this alter column type"
___
*[pkg/sql/alter_column_type.go, line 135 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/46933#-M7etPfU-8xpgWvkSclx:-M7etPfU-8xpgWvkScly:b9ro05g) ([raw file](https://github.com/cockroachdb/cockroach/blob/7df49a0ffbaba8886de54804adcf69e939c14ba9/pkg/sql/alter_column_type.go#L135)):*
> ```Go
> 	if !params.SessionData().AlterColumnTypeGeneral {
> 		return pgerror.Newf(pgcode.FeatureNotSupported,
> 			"alter column type general is experimental; "+
> ```

I don't think users will understand what "alter column type general" means. How about:

"ALTER TYPE %v is only supported experimentally". And the call to action should go in a hint. Check out how the temp tables one is done in create_table.go
___
*[pkg/sql/alter_column_type.go, line 139 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/46933#-M7etzCeAdTW6u4BD0EJ:-M7etzCeAdTW6u4BD0EK:byeptbc) ([raw file](https://github.com/cockroachdb/cockroach/blob/7df49a0ffbaba8886de54804adcf69e939c14ba9/pkg/sql/alter_column_type.go#L139)):*
> ```Go
> 	}
> 	// Disallow ALTER COLUMN ... TYPE ... USING EXPRESSION.
> 	// Todo(richardjcai): Todo, need to handle "inverse" expression
> ```

nit: Todos should be of the form TODO(richardjcai):
___
*[pkg/sql/alter_column_type.go, line 164 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/46933#-M7euRj2E74klc3m6Epk:-M7euRj2E74klc3m6Epl:b-q6ml76) ([raw file](https://github.com/cockroachdb/cockroach/blob/7df49a0ffbaba8886de54804adcf69e939c14ba9/pkg/sql/alter_column_type.go#L164)):*
> ```Go
> 	// Disallow ALTER COLUMN TYPE general for columns that are
> 	// part of indexes.
> 	for i := range tableDesc.Indexes {
> ```

Use AllNonDropIndexes() to avoid repeating code for the indexes and prmary index.
___
*[pkg/sql/alter_column_type.go, line 226 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/46933#-M7eudJw1tphj8JyyoMV:-M7eudJw1tphj8JyyoMW:b-4b89ur) ([raw file](https://github.com/cockroachdb/cockroach/blob/7df49a0ffbaba8886de54804adcf69e939c14ba9/pkg/sql/alter_column_type.go#L226)):*
> ```Go
> 				return err
> 			}
> 			newDefaultComputedExpr := tree.CastExpr{Expr: expr, Type: t.ToType, SyntaxMode: tree.CastShort}
> ```

It would be better to evaluate the cast expression up front, so it doesn't have to get evaluated once per row at backfill time. You can run .Eval() on it, since default expressions are always constant.
___
*[pkg/sql/alter_column_type_test.go, line 44 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/46933#-M7euums2rM8mFseMLN_:-M7euums2rM8mFseMLNa:b-7mu5ev) ([raw file](https://github.com/cockroachdb/cockroach/blob/7df49a0ffbaba8886de54804adcf69e939c14ba9/pkg/sql/alter_column_type_test.go#L44)):*
> ```Go
> 			RunBeforeChildJobs: func() {
> 				doOnce.Do(func() {
> 					fmt.Println("Once")
> ```

nit: This seems like a debug print. 
___
*[pkg/sql/exec_util.go, line 229 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/46933#-M7ev68q7E3Pq-58WlZD:-M7ev68q7E3Pq-58WlZE:b1m1v5g) ([raw file](https://github.com/cockroachdb/cockroach/blob/7df49a0ffbaba8886de54804adcf69e939c14ba9/pkg/sql/exec_util.go#L229)):*
> ```Go
> 
> var experimentalAlterColumnTypeGeneralMode = settings.RegisterBoolSetting(
> 	"sql.defaults.experimental_experimental_alter_column_type.enabled",
> ```

This has two experimentals in it.
___
*[pkg/sql/exec_util.go, line 230 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/46933#-M7ev7cJ1Sz45QsTpFOW:-M7ev7cJ1Sz45QsTpFOX:b-dg748i) ([raw file](https://github.com/cockroachdb/cockroach/blob/7df49a0ffbaba8886de54804adcf69e939c14ba9/pkg/sql/exec_util.go#L230)):*
> ```Go
> var experimentalAlterColumnTypeGeneralMode = settings.RegisterBoolSetting(
> 	"sql.defaults.experimental_experimental_alter_column_type.enabled",
> 	"default value for experimental_alter_column_type session setting; enables a the use of Alter Column Type for general conversions",
> ```

This should say "enables the use of ALTER COLUMN TYPE for general conversions"
___
*[pkg/sql/logictest/testdata/logic_test/alter_column_type, line 338 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/46933#-M7ewI-aEuubR3G7muc_:-M7ewI-bAVsIPDBi4B0P:bhhfutb) ([raw file](https://github.com/cockroachdb/cockroach/blob/7df49a0ffbaba8886de54804adcf69e939c14ba9/pkg/sql/logictest/testdata/logic_test/alter_column_type#L338)):*
> ```
> CREATE TABLE t12 (x INT check (x > 0))
> 
> statement error pq: unimplemented: alter column type for a column that has a constraint is currently not supported
> ```

Can you add some tests for:

1. changing a column type that's included in the STORING columns of a secondary index
2. changing a column type that's part of a column an an interleaved table
3. collated strings?
___
*[pkg/sql/sessiondata/session_data.go, line 127 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/46933#-M7ewxhd0UyOJ-fOB0vj:-M7ewxhd0UyOJ-fOB0vk:bhcfgdt) ([raw file](https://github.com/cockroachdb/cockroach/blob/7df49a0ffbaba8886de54804adcf69e939c14ba9/pkg/sql/sessiondata/session_data.go#L127)):*
> ```Go
> 	// AlterColumnTypeGeneral is true if ALTER TABLE ... ALTER COLUMN ... TYPE x
> 	// may be used for general conversions requiring online schema change/
> 	AlterColumnTypeGeneral bool
> ```

add Enabled to the end of this


<!-- Sent from Reviewable.io -->

thoszhang reviewed (commented):
@richardjcai Does this allow ALTER COLUMN TYPE to occur in the same transaction as other schema changes? It looks like you can't do it after other mutations have been queued, but can you queue other mutations after a column swap mutation has been queued? We should have tests for both of these cases.

@rohany In an attempt to mark the >100 Github-originating comments as resolved in Reviewable, I clicked the button to "dismiss" you from the conversation. I think this did the right thing in the Reviewable UI, but let me know if this caused problems.

> Reviewed 1 of 26 files at r14, 1 of 17 files at r15, 3 of 17 files at r16, 17 of 22 files at r17, 4 of 4 files at r18.
> Dismissed @rohany from 110 discussions.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46933#-:-M7f2ikqAvwhAD1xC7S5:b-5eamb4)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RichardJCai)

___
*[pkg/sql/alter_column_type.go, line 108 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/46933#-M7ek9-U9IMxEXFlyHrp:-M7ek9-U9IMxEXFlyHrq:bsjs4bi) ([raw file](https://github.com/cockroachdb/cockroach/blob/7df49a0ffbaba8886de54804adcf69e939c14ba9/pkg/sql/alter_column_type.go#L108)):*
> ```Go
> 			return err
> 		}
> 		if err := params.p.createOrUpdateSchemaChangeJob(params.ctx, tableDesc, "alter column type", tableDesc.ClusterVersion.NextMutationID); err != nil {
> ```

This should have the actual `ALTER TABLE` statement.
___
*[pkg/sql/alter_column_type.go, line 113 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/46933#-M7ekU-m-0XzylwuIGea:-M7ekU-m-0XzylwuIGeb:b-3epjg2) ([raw file](https://github.com/cockroachdb/cockroach/blob/7df49a0ffbaba8886de54804adcf69e939c14ba9/pkg/sql/alter_column_type.go#L113)):*
> ```Go
> 		params.p.SendClientNotice(params.ctx, pgnotice.Newf("alter column type changes are finalized asynchronously; "+
> 			"further schema changes on this table may be restricted until the job completes; "+
> 			"some inserts into the altered column may be rejected until the schema change is finalized"))
> ```

nit: `some writes to the altered column`?
___
*[pkg/sql/alter_column_type.go, line 115 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/46933#-M7ekZcEDvGPspWHI0DK:-M7ekZcEDvGPspWHI0DL:byze5nq) ([raw file](https://github.com/cockroachdb/cockroach/blob/7df49a0ffbaba8886de54804adcf69e939c14ba9/pkg/sql/alter_column_type.go#L115)):*
> ```Go
> 			"some inserts into the altered column may be rejected until the schema change is finalized"))
> 	default:
> 		return fmt.Errorf("unknown conversion for %s -> %s",
> ```

nit: This should be an assertion error, right?
___
*[pkg/sql/alter_column_type.go, line 131 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/46933#-M7ekhCz-YRUYW44T38E:-M7ekhCz-YRUYW44T38F:bzhsgxq) ([raw file](https://github.com/cockroachdb/cockroach/blob/7df49a0ffbaba8886de54804adcf69e939c14ba9/pkg/sql/alter_column_type.go#L131)):*
> ```Go
> 	if !params.p.ExecCfg().Settings.Version.IsActive(params.ctx, clusterversion.VersionAlterColumnTypeGeneral) {
> 		return pgerror.Newf(pgcode.FeatureNotSupported,
> 			"all nodes are not the correct version for alter column type general")
> ```

nit: `not all nodes`?
___
*[pkg/sql/alter_column_type.go, line 151 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/46933#-M7elvde5x1i7k8BnC6Y:-M7elvde5x1i7k8BnC6Z:bqkfok1) ([raw file](https://github.com/cockroachdb/cockroach/blob/7df49a0ffbaba8886de54804adcf69e939c14ba9/pkg/sql/alter_column_type.go#L151)):*
> ```Go
> 	}
> 
> 	// Disallow ALTER COLUMN TYPE general for columns that have a constraint.
> ```

We should also check for foreign keys involving this column, right?
___
*[pkg/sql/backfill.go, line 1359 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/46933#-M7esCV36Qf3eSiAQ4HU:-M7esCV36Qf3eSiAQ4HV:b-1td9t6) ([raw file](https://github.com/cockroachdb/cockroach/blob/7df49a0ffbaba8886de54804adcf69e939c14ba9/pkg/sql/backfill.go#L1359)):*
> ```Go
> 			case *sqlbase.DescriptorMutation_ComputedColumnSwap:
> 				return unimplemented.NewWithIssue(47537,
> 					"performing an ALTER TABLE ... ALTER COLUMN TYPE in the same txn as the table was created is not supported")
> ```

nit: `in the same transaction in which the table was created`
___
*[pkg/sql/err_count_test.go, line 66 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/46933#-M7esNZS6oBrnfwd4gLq:-M7esNZS6oBrnfwd4gLr:bttz8yy) ([raw file](https://github.com/cockroachdb/cockroach/blob/d244a6d91a259c256526e2dedd030bb604ef4d59/pkg/sql/err_count_test.go#L66)):*
> ```Go
> }
> 
> func TestUnimplementedCounts(t *testing.T) {
> ```

Is this test actually specific to this feature, or is it meant to be for testing telemetry for unimplemented features in general? If the latter, we should keep this test and replace it with a different unimplemented feature.
___
*[pkg/sql/schema_changer.go, line 890 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/46933#-M7f017y1MTFQmB76xg-:-M7f017y1MTFQmB76xg0:b-3thdls) ([raw file](https://github.com/cockroachdb/cockroach/blob/7df49a0ffbaba8886de54804adcf69e939c14ba9/pkg/sql/schema_changer.go#L890)):*
> ```Go
> 				}
> 				// If we performed MakeMutationComplete on a PrimaryKeySwap mutation, then we need to start
> 				// a job for the index deletion mutations that the primary key swap mutation added, if any.
> ```

This comment should also mention the column swap.
___
*[pkg/sql/logictest/testdata/logic_test/alter_column_type, line 280 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/46933#-M7f14Lh2R3jEmacE6dB:-M7f14Lh2R3jEmacE6dC:b-9lhwf1) ([raw file](https://github.com/cockroachdb/cockroach/blob/7df49a0ffbaba8886de54804adcf69e939c14ba9/pkg/sql/logictest/testdata/logic_test/alter_column_type#L280)):*
> ```
>             )
> 
> # Ensure default expressions are casted correctly.
> ```

Can you add a test for what happens if evaluating the new default expression fails?


<!-- Sent from Reviewable.io -->

RichardJCai commented:
I've addressed the PR review comments by Jordan and Lucy.

@lucy-zhang 
> func TestUnimplementedCounts(t *testing.T) {
> Is this test actually specific to this feature, or is it meant to be for testing telemetry for > unimplemented features in general? If the latter, we should keep this test and replace it with a different unimplemented feature.

This test seems to be testing unimplemented features in general, however I think it this is a somewhat flawed test if we have to replace it whenever we implement the feature that's being used to test it. Also there are telemetry logictests that cover this. 

>@RichardJCai Does this allow ALTER COLUMN TYPE to occur in the same transaction as other schema changes? It looks like you can't do it after other mutations have been queued, but can you queue other mutations after a column swap mutation has been queued? We should have tests for both of these cases.

I've also disabled this ALTER COLUMN TYPE in explicit transactions in general since it's pretty hard to handle without transactional schema changes. After performing ALTER COLUMN TYPE in a transaction, any schema change that references the column will reference the old column, ie a constraint will be added to the old column, then the old column will be dropped and the constraint will apply to a dropped column after the swap. Created this issue to track this https://github.com/cockroachdb/cockroach/issues/49351.

thoszhang reviewed (commented):
> This test seems to be testing unimplemented features in general, however I think it this is a somewhat flawed test if we have to replace it whenever we implement the feature that's being used to test it. Also there are telemetry logictests that cover this.

That may be true, but I'd rather not remove the test in this PR.

> I've also disabled this ALTER COLUMN TYPE in explicit transactions in general since it's pretty hard to handle without transactional schema changes. After performing ALTER COLUMN TYPE in a transaction, any schema change that references the column will reference the old column, ie a constraint will be added to the old column, then the old column will be dropped and the constraint will apply to a dropped column after the swap. Created this issue to track this #49351.

That makes sense. What about something like `alter table t alter column a set not null, alter column a set data type int;`? 

> Reviewed 15 of 15 files at r19.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46933#-:-M8DFeC52mnF2lAv1TEU:bbn496l)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RichardJCai)

___
*[pkg/clusterversion/cockroach_versions.go, line 498 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/46933#-M8DCh2G70NoqrGxA1Fh:-M8DCh2G70NoqrGxA1Fi:b-ghkjdm) ([raw file](https://github.com/cockroachdb/cockroach/blob/7705d9f0cfcf62d67d4f0201cdf3c46324ac85e4/pkg/clusterversion/cockroach_versions.go#L498)):*
> ```Go
> 	{
> 		// VersionAlterColumnTypeGeneral enables the use of alter column type for
> 		// non-trivial conversions.
> ```

I might actually say `conversions that require the column data to be rewritten` or something.
___
*[pkg/sql/alter_column_type.go, line 108 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/46933#-M7ek9-U9IMxEXFlyHrp:-M8DDNBHDOw3VubmacF9:b-rckdyo) ([raw file](https://github.com/cockroachdb/cockroach/blob/7df49a0ffbaba8886de54804adcf69e939c14ba9/pkg/sql/alter_column_type.go#L108)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

This should have the actual `ALTER TABLE` statement.
</blockquote></details>

`tree.AsStringWithFQNames(t, params.Ann())`
___
*[pkg/sql/alter_column_type_test.go, line 214 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/46933#-M8DFQYX9TZ8qsVKzwbL:-M8DFQYY28-8UMcCSpJH:b-7guxaz) ([raw file](https://github.com/cockroachdb/cockroach/blob/7df49a0ffbaba8886de54804adcf69e939c14ba9/pkg/sql/alter_column_type_test.go#L214)):*
> ```Go
> }
> 
> func TestSchemaChangeBeforeAlterColumnType(t *testing.T) {
> ```

Did this get turned into a different test?
___
*[pkg/sql/backfill.go, line 1359 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/46933#-M7esCV36Qf3eSiAQ4HU:-M8DEhpS3erUrVDDE27R:b319dkb) ([raw file](https://github.com/cockroachdb/cockroach/blob/7df49a0ffbaba8886de54804adcf69e939c14ba9/pkg/sql/backfill.go#L1359)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

nit: `in the same transaction in which the table was created`
</blockquote></details>

That's not quite the same as what you have.


<!-- Sent from Reviewable.io -->

RichardJCai commented:
> That makes sense. What about something like `alter table t alter column a set not null, alter column a set data type int;`?
> 
Just to clarify, are you saying this should be allowed? The case where the ALTER COLUMN TYPE is the last schema change in the transaction is alright but to do that, we would have to go and block schema changes affecting the column after the ALTER COLUMN TYPE in the transaction. I think this should be done in a different PR so we don't add to much more code to this one. What are your thoughts?

RichardJCai commented:
> ```go
> ```go
> func TestSchemaChangeBeforeAlterColumnType(t *testing.T) {
> ```
> ```

Also this became a logic test since it just became testing that ALTER COLUMN TYPE 


> > This test seems to be testing unimplemented features in general, however I think it this is a somewhat flawed test if we have to replace it whenever we implement the feature that's being used to test it. Also there are telemetry logictests that cover this.
> 
> That may be true, but I'd rather not remove the test in this PR.
> 
> > I've also disabled this ALTER COLUMN TYPE in explicit transactions in general since it's pretty hard to handle without transactional schema changes. After performing ALTER COLUMN TYPE in a transaction, any schema change that references the column will reference the old column, ie a constraint will be added to the old column, then the old column will be dropped and the constraint will apply to a dropped column after the swap. Created this issue to track this #49351.
> 
> That makes sense. What about something like `alter table t alter column a set not null, alter column a set data type int;`?
> 
> > Reviewed 15 of 15 files at r19.
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46933#-:-M8DFeC52mnF2lAv1TEU:bbn496l)_ status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RichardJCai)
> 
> _[pkg/clusterversion/cockroach_versions.go, line 498 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/46933#-M8DCh2G70NoqrGxA1Fh:-M8DCh2G70NoqrGxA1Fi:b-ghkjdm) ([raw file](https://github.com/cockroachdb/cockroach/blob/7705d9f0cfcf62d67d4f0201cdf3c46324ac85e4/pkg/clusterversion/cockroach_versions.go#L498)):_
> 
> > ```go
> > 	{
> > 		// VersionAlterColumnTypeGeneral enables the use of alter column type for
> > 		// non-trivial conversions.
> > ```
> 
> I might actually say `conversions that require the column data to be rewritten` or something.
> 
> _[pkg/sql/alter_column_type.go, line 108 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/46933#-M7ek9-U9IMxEXFlyHrp:-M8DDNBHDOw3VubmacF9:b-rckdyo) ([raw file](https://github.com/cockroachdb/cockroach/blob/7df49a0ffbaba8886de54804adcf69e939c14ba9/pkg/sql/alter_column_type.go#L108)):_
> 
> Previously, lucy-zhang (Lucy Zhang) wrote…
> > This should have the actual `ALTER TABLE` statement.
> 
> `tree.AsStringWithFQNames(t, params.Ann())`
> 
> _[pkg/sql/alter_column_type_test.go, line 214 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/46933#-M8DFQYX9TZ8qsVKzwbL:-M8DFQYY28-8UMcCSpJH:b-7guxaz) ([raw file](https://github.com/cockroachdb/cockroach/blob/7df49a0ffbaba8886de54804adcf69e939c14ba9/pkg/sql/alter_column_type_test.go#L214)):_
> 
> > ```go
> > }
> > 
> > func TestSchemaChangeBeforeAlterColumnType(t *testing.T) {
> > ```
> 
> Did this get turned into a different test?
> 
> _[pkg/sql/backfill.go, line 1359 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/46933#-M7esCV36Qf3eSiAQ4HU:-M8DEhpS3erUrVDDE27R:b319dkb) ([raw file](https://github.com/cockroachdb/cockroach/blob/7df49a0ffbaba8886de54804adcf69e939c14ba9/pkg/sql/backfill.go#L1359)):_
> 
> Previously, lucy-zhang (Lucy Zhang) wrote…
> > nit: `in the same transaction in which the table was created`
> 
> That's not quite the same as what you have.

Re-added back the unimplemented test and TestSchemaChangeBeforeAlterColumnType test (wasn't supposed to remove the last one). 

And addressed the other comments.

thoszhang reviewed (approved):
> Reviewed 5 of 5 files at r20, 5 of 5 files at r21.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46933)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RichardJCai)

___
*[pkg/sql/alter_column_type.go, line 49 at r21](https://reviewable.io/reviews/cockroachdb/cockroach/46933#-M8M4U7r77xyzyoSJtso:-M8M4U7s6i6N0J7YxtmH:bi0hwd0) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3e0c75ea3c61e720589c756bf65c9a10576e807/pkg/sql/alter_column_type.go#L49)):*
> ```Go
> var alterColTypeInCombinationNotSupportedErr = unimplemented.NewWithIssuef(
> 	49351, "ALTER COLUMN TYPE cannot be used in combination "+
> 		"with other ALTER TABLE statements")
> ```

This should be `commands` instead of statements


<!-- Sent from Reviewable.io -->

RichardJCai commented:
bors r+

RichardJCai commented:
Thanks for the reviews @rohany @lucy-zhang @jordanlewis!

ericharmeling commented:
@RichardJCai 
Is the plan to keep the `enable_experimental_alter_column_type_general` setting for 20.2 GA?

RichardJCai commented:
> @RichardJCai
> Is the plan to keep the `enable_experimental_alter_column_type_general` setting for 20.2 GA?

Yeah that's the plan

knz commented:
Would it be possible to drop the "experimental" word from the variable name, and only highlight its experimental nature in the description string and in docs?

That will ensure that later users don't need to change their app code when we promote the feature to non experimental..

-- 
Verstuurd vanaf mijn Android apparaat met K-9 Mail. Excuseer mijn beknoptheid.
